external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -166,30 +183,84 @@ private[sql] class AvroSerializer(
         (getter, ordinal) => ByteBuffer.wrap(getter.getBinary(ordinal))
 
       case (DateType, INT) =>
-        (getter, ordinal) => dateRebaseFunc(getter.getInt(ordinal))
+        (getter, ordinal) =>
+          getter.get(ordinal, DateType) match {
```

Comment:
Nit: we can save the Int result to a variable and then call `dateRebaseFunc(..)` at the end to reduce duplicated code.